transcendence-app/src/game/game.queue.gateway.ts

```diff
@@ -0,0 +1,160 @@
+import {
+  ClassSerializerInterceptor,
+  UseFilters,
+  UseGuards,
+  UseInterceptors,
+} from '@nestjs/common';
+import {
+  ConnectedSocket,
+  MessageBody,
+  SubscribeMessage,
+  WebSocketGateway,
+  WebSocketServer,
+  WsException,
+} from '@nestjs/websockets';
+import { Server, Socket } from 'socket.io';
+import { BadRequestTransformationFilter } from '../shared/filters/bad-request-transformation.filter';
+import { TwoFactorAuthenticatedGuard } from '../shared/guards/two-factor-authenticated.guard';
+import { GameQueueService } from './game.queue.service';
+import {
+  GameChallengeResponseDto,
+  GameChallengeStatus,
+  gameQueueClientToServerWsEvents,
+  gameQueueServerToClientWsEvents,
+  GameStatus,
+  GameStatusUpdateDto,
+  GameUserChallengeDto,
+} from 'pong-engine';
+import { EventEmitter2 } from '@nestjs/event-emitter';
+import { SocketService } from '../socket/socket.service';
+
+@WebSocketGateway({ path: '/api/v1/socket.io' })
+@UseGuards(TwoFactorAuthenticatedGuard)
+@UseInterceptors(ClassSerializerInterceptor)
+@UseFilters(BadRequestTransformationFilter)
+export class GameQueueGateway {
+  @WebSocketServer() server!: Server;
+  constructor(
+    private gameQueueService: GameQueueService,
+    private eventEmitter: EventEmitter2,
+    private socketService: SocketService,
+  ) {}
+
+  afterInit(server: Server) {
+    this.gameQueueService.socket = server;
+  }
+
+  @SubscribeMessage(gameQueueClientToServerWsEvents.gameQueueJoin)
+  async gameQueueJoin(@ConnectedSocket() client: Socket): Promise<boolean> {
+    const retVal = await this.gameQueueService.gameQueueJoin(
+      client.request.user.id,
+    );
+    if (retVal) {
+      const [gameRoomId, isRoomFull] = retVal;
+      client.join(gameRoomId);
+      if (isRoomFull) {
+        this.server
+          .to(gameRoomId)
+          .emit(gameQueueServerToClientWsEvents.gameStatusUpdate, {
+            status: GameStatus.READY,
+            gameRoomId,
+          } as GameStatusUpdateDto);
+        this.eventEmitter.emit('game.ready', {
+          status: GameStatus.READY,
+          gameRoomId,
+        } as GameStatusUpdateDto);
```

Comment:
suggestion: Is it possible to also emit the player IDs from here because we need them to create a game?

Reply:
done